include/ginkgo/core/config/registry.hpp

```diff
@@ -0,0 +1,220 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_REGISTRY_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_REGISTRY_HPP_
+
+
+#include <complex>
+#include <functional>
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils_helper.hpp>
+#include <ginkgo/core/config/property_tree.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+class registry;
+
+/**
+ * type_descriptor gives the initial default type for common type in ginkgo such
+ * as ValueType and IndexType. If the factory config does not specify these
+ * type, the configuration will use this as the default.
+ */
+class type_descriptor {
```

Comment:
In the property tree, it was requested to put two structures to one file (not merge structure).
I do not want to move it in/out repeatedly.

Reply:
IMO that was different, there it was about a struct that's only an implementation detail and the user should never interface with. Here this is a class that the user should use. Thus I think it deserves it's own file. Besides, it's used in other places than the registry, so that's another argument to split it up.